electrum/lnpeer.py

```diff
@@ -1769,29 +1769,69 @@ async def _shutdown(self, chan: Channel, payload, *, is_local: bool):
-        our_fee = fee_rate * closing_tx.estimated_size() // 1000
+        our_fee = fee_rate * closing_tx.estimated_size() // 1000  # inclusion in one block
+        # TODO: anchors: remove this, as commitment fee rate can be below chain head fee rate?
         # BOLT2: The sending node MUST set fee less than or equal to the base fee of the final ctx
         max_fee = chan.get_latest_fee(LOCAL if is_local else REMOTE)
         our_fee = min(our_fee, max_fee)
-        drop_to_remote = False
+
+        drop_to_remote = False  # does the peer drop its to_local output or not?
         def send_closing_signed():
+            MODERN_FEE = True
+            if MODERN_FEE:
+                nonlocal fee_range_sent  # we change fee_range_sent in outer scope
+                fee_range_sent = fee_range
+                closing_signed_tlvs = {'fee_range': fee_range}
+            else:
+                closing_signed_tlvs = {}
+
             our_sig, closing_tx = chan.make_closing_tx(our_scriptpubkey, their_scriptpubkey, fee_sat=our_fee, drop_remote=drop_to_remote)
-            self.send_message('closing_signed', channel_id=chan.channel_id, fee_satoshis=our_fee, signature=our_sig)
+            self.logger.info(f"Sending fee range: {closing_signed_tlvs} and fee: {our_fee}")
+            self.send_message(
+                'closing_signed',
+                channel_id=chan.channel_id,
+                fee_satoshis=our_fee,
+                signature=our_sig,
+                closing_signed_tlvs=closing_signed_tlvs,
+            )
         def verify_signature(tx, sig):
             their_pubkey = chan.config[REMOTE].multisig_key.pubkey
             preimage_hex = tx.serialize_preimage(0)
             pre_hash = sha256d(bfh(preimage_hex))
             return ecc.verify_signature(their_pubkey, sig, pre_hash)
+
+        # this is the fee range we initially try to enforce
+        # we aim at a fee between next block inclusion and some lower value
+        fee_range = {'min_fee_satoshis': our_fee // 2, 'max_fee_satoshis': our_fee}
+        their_fee = None
+        fee_range_sent = {}
+        is_initiator = chan.constraints.is_initiator
         # the funder sends the first 'closing_signed' message
-        if chan.constraints.is_initiator:
+        if is_initiator:
             send_closing_signed()
+
         # negotiate fee
         while True:
-            # FIXME: the remote SHOULD send closing_signed, but some don't.
-            cs_payload = await self.wait_for_message('closing_signed', chan.channel_id)
+            try:
+                cs_payload = await self.wait_for_message('closing_signed', chan.channel_id)
+            except asyncio.exceptions.TimeoutError:
+                if not is_initiator and not their_fee:  # we only force close if a peer doesn't reply
+                    await self.lnworker.force_close_channel(chan.channel_id)
+                    raise Exception("Peer didn't reply with closing signed, force closed.")
```

Comment:
Note: the BOLT does not say that we have to force close. 
I suppose this choice follows from https://github.com/lightning/bolts/pull/847/files#r689977518

Reply:
Note 2: `_shutdown` is called from two separate places, one of them is is a `Taskgroup`.
In that case, `_shutdown` could be cancelled if the remote closes the connection, and might not have time to force close.

This way of calling a task from non-homogeneous contexts seems dangerous. This comment is not about this specific pull request, but about how we expect code to be executed. A similar issue was encountered in `request_force_close_from_backup`

Reply:
I see, but I also don't recall why I added the force close here, what would you suggest?

Reply:
Sorry, I had not seen the new definition of "fail the channel". Indeed, we have to force close.
Maybe this should be using try_force_closing, as in #7645.

Reply:
we don't need to wait for #7645. See the new API, `schedule_force_closing` in 556b98736e78eb04bc68fd44e701e3fa174023a9

Reply:
Also, the BOLT does not make a distinction between funder and non-funder. I think we have to force close in both cases.

Reply:
> we don't need to wait for #7645. See the new API, `schedule_force_closing` in [556b987](https://github.com/spesmilo/electrum/commit/556b98736e78eb04bc68fd44e701e3fa174023a9)

I'll use it, but I meant that we introduce a wrapper `lnpeer.schedule_force_closing` that is added and could be used here as well.

> Also, the BOLT does not make a distinction between funder and non-funder. I think we have to force close in both cases.

My motivation was to catch the case where we would have sent an initial `closing_signed` message (as the initiator) and wait for the first response of the peer, if it doesn't, force close. I haven't tackled the case when the peer doesn't reply during a general step of the negotiation, of course we could close there as well.